### PR TITLE
Eliminate unused virtual methods we detect after optimizations

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Before this PR, it was not possible to get rid of unused virtual method slots once whole program analysis determined they are needed. This PR makes it (sort of) possible.

Codegen needs to know what numerical slots get assigned to virtual methods because we use the slots to index into vtable to do virtual calls/generic lookup. We also want to be able to skip generating unused virtual methods. Before we do actual codegen optimizations, we don't know what virtual method uses can actually be optimized away.

### Before this PR

Without optimizations: don’t do codegen that needs slots inlined into code generated by RyuJIT. Instead instruct RyuJIT to call a helper. The helper is emitted as assembly by the compiler driver itself at the time all used virtual methods are known and we can stabilize slot numbers. When assigning slots, we skip the virtual methods that are unused.

With optimizations: assign slots after whole program analysis is done and we know what slots can be skipped. Codegen might still do optimizations that avoids actually needing a slot. We still generate the slot and fill it out and it’s a bit of a waste.

### After this PR

Without optimizations: assign slots based on what we see in metadata. Do codegen based on this information and allow inlining slot indices into codegen. Still try to find unused slots. Generate unused slots as null slots (place a literal null pointer in the vtable). This is a both an improvement and a regression. It’s debug so it doesn’t matter much. We can delete the handwritten assembly, so it’s an improvement there.

With optimizations: assign slots after whole program analysis. Still try to find unused slots. If an unused slot is found, generate it as null. We can’t eliminate it anymore for regular vtable slots (because we did codegen that assumes indices into vtable), but at least we don’t have to generate a standalone method body and a reloc to it. For sealed vtable slots, we can actually eliminate the slot (the slot is not inlined in code) and this PR does that.

Cc @dotnet/ilc-contrib

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
# Size statistics

Pull request dotnet/runtime#100286

| Project | Size before | Size after | Difference |
|---------|-------------|------------|------------|
| avalonia.app-linux | [ 24544208 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359332182) | [ 24389968 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359332176) | -154240 |
| avalonia.app-windows | [ 22196224 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359334394) | [ 22054912 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359335832) | -141312 |
| hello-linux | [ 1266592 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359328576) | [ 1254192 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359328939) | -12400 |
| hello-minimal-linux | [ 1118984 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359328812) | [ 1110712 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359328797) | -8272 |
| hello-minimal-windows | [ 898048 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359332329) | [ 883712 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359331665) | -14336 |
| hello-windows | [ 1151488 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359331694) | [ 1133056 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359333245) | -18432 |
| webapiaot-linux | [ 10268512 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359330404) | [ 10187936 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359330621) | -80576 |
| webapiaot-windows | [ 9346048 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359332891) | [ 9281024 ](https://github.com/MichalStrehovsky/rt-sz/actions/runs/8436439868/artifacts/1359332445) | -65024 |

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs [hunk 1]
```diff
@@ -273,7 +273,16 @@ public ScannedVTableProvider(ImmutableArray<DependencyNodeCore<NodeFactory>> mar
                     var vtableSliceNode = node as VTableSliceNode;
                     if (vtableSliceNode != null)
                     {
-                        _vtableSlices.Add(vtableSliceNode.Type, vtableSliceNode.Slots);
+                        var usedSlots = default(ArrayBuilder<MethodDesc>);
```
Comment (am11) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:276]: nodiff style nit:
```suggestion
                        ArrayBuilder<MethodDesc> usedSlots = new();
```

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:276]: https://github.com/dotnet/runtime/blob/b432fc66fe59873b2b2ad584c04eaf1e4b11d611/eng/CodeAnalysis.src.globalconfig#L1280-L1281

Comment (am11) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:276]: Fixed.

File: src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs [hunk 2]
```diff
@@ -1615,31 +1615,14 @@ private void getCallInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_RESO
 
                 pResult->nullInstanceCheck = false;
             }
-            else if ((flags & CORINFO_CALLINFO_FLAGS.CORINFO_CALLINFO_LDFTN) == 0
-                // Canonically-equivalent types have the same vtable layout. Check the canonical form.
-                // We don't want to accidentally ask about Foo<object, __Canon> that may or may not
-                // be available to ask vtable questions about.
-                // This can happen in inlining that the scanner didn't expect.
-                && _compilation.HasFixedSlotVTable(targetMethod.OwningType.ConvertToCanonForm(CanonicalFormKind.Specific)))
+            else if ((flags & CORINFO_CALLINFO_FLAGS.CORINFO_CALLINFO_LDFTN) == 0)
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs:1618]: Is there a reason for suppressing VTABLE slot for CORINFO_VIRTUALCALL_VTABLE? We should be able to take it to fetch the virtual delegate target too (it is probably not super common).

**Comment by jkotas** [comment]
> We can’t eliminate it anymore for regular vtable slots (because we did codegen that assumes indices into vtable)

If we wanted to fix this too, I guess we can emit vtable indices relocs in codegen and do a pass to shrink the indices to avoid the null entries. It probably would not save much.

**Review by jkotas — APPROVED** [review]
Nice!

**Comment by MichalStrehovsky** [comment]
> > We can’t eliminate it anymore for regular vtable slots (because we did codegen that assumes indices into vtable)
> 
> If we wanted to fix this too, I guess we can emit vtable indices relocs in codegen and do a pass to shrink the indices to avoid the null entries. It probably would not save much.

Yep, but we'd also want different size relocs (fixup for value that is smaller than whatever is efficient instruction encoding for small constants, etc.) so that we don't make virtual call sequences much larger than needed. It might be a small improvement to working set that we're still leaving on the table, but there's not _too many_ of these slots and it's probably not worth the hassle.

File: src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs (hunk 2 above)
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs:1618]: I'll look into this once we get runtime tests testing in order. As you say, it's not super common - I think we'd only exercise the results of this when loading addresses of virtual methods in situations that are not verifiable-delegate-sequence. Those tests are in the runtime test tree.

**Pull request merged**